### ISSUE TITLE
Merge Dataframe trait and DataframeImpl into Dataframe struct

### ISSUE THREAD

**Issue by houqp** [apache/arrow-datafusion#1962]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Follow up for https://github.com/apache/arrow-datafusion/pull/1712#issuecomment-1025266914.

The Dataframe trait was introduced in order to have two separate Dataframe implementation in Datafusion and Ballista. Since then, the design has changed and we are sharing the same dataframe implementation between Datafusion and Ballista now. Therefore, we could simplify the code base by getting rid of the Dataframe trait abstraction and promoting DataframeImpl into Dataframe struct.

**Describe the solution you'd like**

Dataframe trait and DataframeImpl merged into a single Dataframe struct

**Additional context**

Should help unblock https://github.com/apache/arrow-datafusion/pull/1712
